src/openforms/submissions/api/validation.py

```diff
@@ -71,7 +71,10 @@ def _run_formio_validation(self) -> None:
                 step_data_serializer = build_serializer(
                     configuration["components"],
                     data=data,
-                    context={"submission": submission},
+                    context={
+                        "submission": submission,
+                        "request": self.context["request"],
+                    },
```

Comment:
I'm not sure what's wise here - this addresses the problem indeed, but it does tie formio validation to a request-response cycle and as you noticed, it makes test code not particularly elegant.

I think I wrote down somewhere that IMO there's no reason why the uploads need to be tracked individually in the session - I think we should be able to:

* send the submission UUID/URL during the file upload in the SDK
* link the temporary upload to the submission record. the security check here then only needs to validate ownership of the submission
* when validating, we already have access to the submission instance, so we should only consider temporary uploads within that submission

This has a number of benefits:

* less session polution (and much nicer: race conditions when updating the array of values!)
* all information is contained within a submission, which doesn't require a request/response cycle for processing (that includes validation)
* when submissions are pruned, temporary uploads still left over (because of whatever reason) can be cascade-deleted, so we don't need the periodic cleanup of temporary uploads anymore

Reply:
Uploads are created before submission steps are.